[mlir][emitc] Add address-of and dereference ops

aniragil opened the pull request:
EmitC currently models C's `&` and `*` operators via its `apply` op, which has several drawbacks:
    
- Representing multiple opcodes (selected by an attribute) in a single op complicates the code by adding a second, attribute-based selection layer on top of MLIR's standard isa<> mechanism.
    
- Its pre-lvalue semantics combines dereferencing with memory access and allows taking the address of SSA values.
    
This patch adds two distinct ops to model these C operators. These ops are lvalue-based and therefore don't incur any side-effects. EmitC passes were converted to use the new ops instead of `apply` which is now deprecated, to be retired in the future.

aniragil commented:
@simon-camp , @ayalz

marbre reviewed (commented):
This breaks the use of EmitC in IREE without providing a suitable alternative. **Therefore, this cannot be merged as-is!**
In IREE an address-of operator is applied to the arguments passed to a function (see [ConvertVMToEmitC.cpp#L2155](https://github.com/openxla/iree/blob/7963ca781b6f3782ad21d20ac24b7adca150e36f/compiler/src/iree/compiler/Dialect/VM/Conversion/VMToEmitC/ConvertVMToEmitC.cpp#L2155)) and the &-operator is also used in conjunction with IREE's `!vm.ref` type.

Furthermore, I don't see a strong argument so far to split up the `emitc.apply` op. In the initial proposal, presented at the ODM at 2021-04-22 (EmitC: Generating C/C++ from MLIR [slides](https://drive.google.com/file/d/1p2AM7B1beb5oVf_tlTyinnVQoWuodoxl/view?usp=sharing)), we had an `emitc.getaddressof` op, but it was later on agreed to rather go with the `emitc.apply` op.

From my point of view it is much more important to answer the question, whether it should be ensured in the conversion that the &-operator is not used incorrectly or if an incorrect usage could be prevented with a verification pass instead of coupling the &-operator to only EmitC variables.

aniragil commented:
> This breaks the use of EmitC in IREE without providing a suitable alternative. **Therefore, this cannot be merged as-is!** In IREE an address-of operator is applied to the arguments passed to a function (see [ConvertVMToEmitC.cpp#L2155](https://github.com/openxla/iree/blob/7963ca781b6f3782ad21d20ac24b7adca150e36f/compiler/src/iree/compiler/Dialect/VM/Conversion/VMToEmitC/ConvertVMToEmitC.cpp#L2155)) and the &-operator is also used in conjunction with IREE's `!vm.ref` type.
> 

Sorry, should have described it in the commit message: The alternative to taking the address of a value is to define a variable using emitc::variable, use an emitc::assign to assign the value to that variable and then take that variable's address using emitc::address_of. In the mentioned iree code this seems to require a rather minor change in [the addressOf function](https://github.com/openxla/iree/blob/7963ca781b6f3782ad21d20ac24b7adca150e36f/compiler/src/iree/compiler/Dialect/VM/Conversion/VMToEmitC/EmitCBuilders.cpp#L132), but I'm not familiar with iree's code so I may be completely off here.

> Furthermore, I don't see a strong argument so far to split up the `emitc.apply` op. In the initial proposal, presented at the ODM at 2021-04-22 (EmitC: Generating C/C++ from MLIR [slides](https://drive.google.com/file/d/1p2AM7B1beb5oVf_tlTyinnVQoWuodoxl/view?usp=sharing)), we had an `emitc.getaddressof` op, but it was later on agreed to rather go with the `emitc.apply` op.
> 

The current state is that emitc:apply as a single op is modelling two distinct opcodes and selecting between them using an attribute, which complicates any code handling this op. I don't know how often this approach is used in MLIR, but I remember  @joker-eph explaining its downsides during the [2021-05-27 MLIR Open Meeting](https://youtu.be/_qXCzP5YAZY?t=3237). However the emitc.apply case might be different. Could you elaborate on why the multi-op approach was eventually chosen and why the same approach wasn't applied to emitc's binary ops?

> From my point of view it is much more important to answer the question, whether it should be ensured in the conversion that the &-operator is not used incorrectly or if an incorrect usage could be prevented with a verification pass instead of coupling the &-operator to only EmitC variables.

This patch suggests a possible way to address the specific concern of breaking IREE by limiting emitc.apply **assuming** the suggestion to separate SSA values from variables is accepted (I'll mark it as draft to avoid confusion), which is a broader discussion yet to be done on discourse. I'll open a topic for it, but in a nutshell - I think limiting address-taking to variables goes beyond verification issues. I agree with @jpienaar that "... separating SSA IDs from variables, variables as symbols and expressions may all be related", as identifying SSA values with locations whose addresses can be taken seems to induce complicated (if not inconsistent) semantics, e.g. does the result of emitc.literal have a location? If the result of emitc.add does, what happens to it when it's part of an expression?

simon-camp commented:
> Sorry, should have described it in the commit message: The alternative to taking the address of a value is to define a variable using emitc::variable, use an emitc::assign to assign the value to that variable and then take that variable's address using emitc::address_of. In the mentioned iree code this seems to require a rather minor change in [the addressOf function](https://github.com/openxla/iree/blob/7963ca781b6f3782ad21d20ac24b7adca150e36f/compiler/src/iree/compiler/Dialect/VM/Conversion/VMToEmitC/EmitCBuilders.cpp#L132), but I'm not familiar with iree's code so I may be completely off here.

I've looked through the uses of the address-of operator in IREE and it is only used on emitc.variables or on function parameters. So the explicit indirection through variables would of course work, but would make the generated code less readable.

> This patch suggests a possible way to address the specific concern of breaking IREE by limiting emitc.apply **assuming** the suggestion to separate SSA values from variables is accepted (I'll mark it as draft to avoid confusion), which is a broader discussion yet to be done on discourse. I'll open a topic for it, but in a nutshell - I think limiting address-taking to variables goes beyond verification issues. I agree with @jpienaar that "... separating SSA IDs from variables, variables as symbols and expressions may all be related", as identifying SSA values with locations whose addresses can be taken seems to induce complicated (if not inconsistent) semantics, e.g. does the result of emitc.literal have a location? If the result of emitc.add does, what happens to it when it's part of an expression?

I agree that the address-of Op should be constrained to C variables (however these may be modeled in the end). However I would argue that function parameters should also be allowed, as by the C standard.
I will work through your RFC in the next days and add my thoughts.

aniragil committed:
[mlir][emitc] Add address-of and dereference ops

EmitC currently models C's `&` and `*` operators via its `apply` op,
which has several drawbacks:

- Representing multiple opcodes (selected by an attribute) in a single
  op complicates the code by adding a second, attribute-based selection
  layer on top of MLIR's standard isa<> mechanism.

- Its pre-lvalue semantics combines dereferencing with memory access and
  allows taking the address of SSA values.

This patch adds two distinct ops to model these C operators. These ops
are lvalue-based and therefore don't incur any side-effects. EmitC
passes were converted to use the new ops instead of `apply` which is
now deprecated, to be retired in the future.

simon-camp commented on mlir/include/mlir/Dialect/EmitC/IR/EmitCTypes.td:
```diff
@@ -185,4 +185,19 @@ class EmitC_LValueOf<list<Type> allowedTypes> :
+> : PredOpTrait<summary, CPred<comparator # "(" #
+      !subst("$_self", "$" # lhsArg # ".getType()", transformL) #
+      ", " #
+      !subst("$_self", "$" # rhsArg # ".getType()", transformL) #
```
I think there is a typo, but I don't see this used currently.
```suggestion
      !subst("$_self", "$" # rhsArg # ".getType()", transformR) #
```

simon-camp reviewed (approved):
Just one nit, overall this looks good.

aniragil committed:
Remove unused trait

aniragil commented:
> Looks good to me. But let's get confirmation from @marbre that his concerns are addressed.

Sure. @marbre, are we good to go on this PR?

Regarding the graceful retirement of `apply`: PR will be followed by a PSA in Discourse encouraging users to switch from the now-deprecated `apply` op to the new ops. For now it remains supported by the translator and form-expressions, but upstream conversion passes don't generate it anymore.

Per the discussion about regarding iree's `addressOf()` - IIUC the suggestion to create a temp variables for arguments and take their address was implemented, so `address_of` can replace `apply` there.

I agree with @simon-camp regarding function arguments being lvalues in C just like variables, so if we'd like to support taking their address in the future I think they should be switched to `lvalue<>` types (means arguments will need to be loaded just like variables, but we can do that transparently when lowering `func::func` to `emitc::func`)

marbre reviewed (commented):
I was ooo during the time this was reviewed and landed but looks good from the discussion here.
